src/Framework/Support/TypedProperties/TypedProperties.php

```diff
@@ -0,0 +1,166 @@
+<?php
+
+namespace Give\Framework\Support\TypedProperties;
+
+use Give\Framework\Exceptions\Primitives\InvalidArgumentException;
+use Give\Framework\Support\TypedProperties\Exceptions\InvalidPropertyType;
+use Give\Framework\Support\TypedProperties\Exceptions\PropertyNotInitialized;
+use Give\Framework\Support\TypedProperties\Exceptions\PropertyReadOnly;
+use Give\Framework\Support\TypedProperties\Exceptions\PropertyWriteOnly;
+
+trait TypedProperties
+{
+    /**
+     * Values may be set as
+     *      "key" => "type"
+     *      "key:readonly"
+     *      "key" => ["type", "default"]
+     *      "key" => ["type:readonly", "default"]
+     *
+     * @var TypedProperty[]|array The properties that are allowed to be set.
+     */
+    protected $properties = [];
```

Comment:
@JasonTheAdams This may be a small thing, but could the name of the `$properties` property be something a bit more specific? A quick code search says we don't have any PHP classes in Give with a `$properties` member (except in `/vendor`), but I'm worried that this trait will produce weird errors when using a class that already has a `$properties` property (as the comment from @alaca indicates.)

Reply:
@LMNTL It would break, apparently, if there's a property conflict. I'm inclined to keep the name, though, as it's not just the definition of the properties, but they ultimately store the property values. I'm not thinking of a more specific or meaningful name. Open to suggestions!